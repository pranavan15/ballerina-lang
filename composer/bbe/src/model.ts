export interface BallerinaExample {
    title?: string;
    url: string;
    name: string;
}

export interface BallerinaExampleCategory {
    title: string;
    column: number;
    samples: Array<BallerinaExample>;
    category?: string;
}  
