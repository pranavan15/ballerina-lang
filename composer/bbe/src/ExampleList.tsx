/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import * as React from 'react';
import { Grid, Input } from 'semantic-ui-react';
import { cloneDeep, debounce } from 'lodash';
import { BallerinaExampleCategory } from './model';

export interface SamplesListState {
    samples? : Array<BallerinaExampleCategory>;
    searchQuery? : string;
}

export interface SamplesListProps {
    openSample: (url: string) => void;
    getSamples: () => Promise<Array<BallerinaExampleCategory>>;
}

/**
 * React component for rendering a list of Ballerina examples.
 *
 * @class SamplesList
 * @extends {Component}
 */
export class SamplesList extends React.Component<SamplesListProps, SamplesListState> {

    private _availableSamples: undefined | Array<BallerinaExampleCategory>;
    private searchInput: null | Input;
    private onSearchQueryEdit: () => void;

    constructor(props: SamplesListProps, context: SamplesListState) {
        super(props, context);
        this.onSearchQueryEdit = debounce(() => {
            const { searchQuery } = this.state;
            if (searchQuery != undefined && this._availableSamples) {
                let samples = cloneDeep(this._availableSamples);
                samples = samples.filter((sampleCategory) => {
                    if (!sampleCategory.title.toLowerCase().includes(searchQuery)) {
                        sampleCategory.samples = sampleCategory.samples.filter(sample => sample.name.toLowerCase().includes(searchQuery));
                    }
                    return sampleCategory.samples.length !== 0;
                });
                this.setState({
                    samples,
                });
            }
        }, 500).bind(this);
    }

    componentDidMount() {
        this.focusOnSearchInput();
        this.props.getSamples().then((samples) => {
            this._availableSamples = samples;
            this.setState({
                samples,
            });
        });
    }

    componentWillReceiveProps(nextProps: SamplesListProps) {
        this.props.getSamples().then((samples) => {
            this._availableSamples = samples;
            this.setState({
                samples,
            });
        });
    }

    focusOnSearchInput() {
        if (this.searchInput) {
            this.searchInput.focus();
        }
    }

    getColumnContents() {

        const columns: Array<Array<BallerinaExampleCategory>> = [];
        const { samples } = this.state;
        if (samples) {
            samples.forEach((sample: BallerinaExampleCategory) => {
                columns[sample.column] = columns[sample.column] || [];
                columns[sample.column].push(sample);
            });
        }
        return columns;
    }

    renderColumnItem(column: BallerinaExampleCategory) {
        return (
            <ul key={column.title}>
                <li className='title'>{column.title}</li>
                <ul>
                    {
                        column.samples.map((sample) => {
                            return (
                            <li className='list-item' key={sample.url}>
                                <a
                                    href='#'
                                    onClick={
                                        () => this.props.openSample(sample.url)}
                                >
                                    {sample.name}
                                </a>
                            </li>);
                        })
                    }
                </ul>
            </ul>
        );
    }

    public render() {
        return (
            <Grid className='welcome-page'>
                <Grid.Row className='welcome-navbar' columns={2}>
                    <Grid.Column className='nav-tagline'>
                        Search and open available examples
                    </Grid.Column>
                    <Grid.Column>
                        <div className='top-nav-links' style={{ paddingRight: 0, marginRight: 40 }}>
                            <Input
                                ref={(ref) => {
                                    this.searchInput = ref;
                                }}
                                loading={!this.state || !this.state.samples}
                                placeholder='Search'
                                onChange={(event: React.SyntheticEvent<HTMLInputElement>) => {
                                    this.setState({
                                        searchQuery: event.currentTarget.value
                                    });
                                    this.onSearchQueryEdit();
                                }}
                            />
                        </div>
                    </Grid.Column>
                </Grid.Row>
                <Grid.Row className='welcome-content-wrapper'>
                    <Grid.Column mobile={16} tablet={16} computer={16} className='rightContainer'>
                        <Grid>
                            {this.state && this.state.samples &&
                                <Grid.Row columns={4} className='sample-wrapper'>
                                    {
                                        this.getColumnContents().map((column, index) => {
                                            return (
                                                <Grid.Column key={index} mobile={16} tablet={8} computer={4}>
                                                    {column.map(columnItem => this.renderColumnItem(columnItem))}
                                                </Grid.Column>
                                            );
                                        })
                                    }
                                </Grid.Row>
                            }
                        </Grid>
                    </Grid.Column>
                </Grid.Row>
            </Grid>);
    }
}
