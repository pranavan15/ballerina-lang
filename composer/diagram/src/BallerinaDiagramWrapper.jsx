/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
import React from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import PropTypes from 'prop-types';
import { Dimmer, Loader } from 'semantic-ui-react';
import DesignView from 'plugins/ballerina/views/design-view';
import TreeBuilder from 'plugins/ballerina/model/tree-builder';
import FragmentUtils from 'plugins/ballerina/utils/fragment-utils';
import '@ballerina/theme/default-theme/index.js/';
import './scss/themes/default.scss';
import '@ballerina/theme/font-ballerina';

const BallerinaDesignView = DragDropContext(HTML5Backend)(DesignView);

const TREE_MODIFIED = 'tree-modified';

/**
 * A wrapper component to wrap Editable Ballerina Diagram.
 */
class BallerinaDiagramWrapper extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            currentAST: undefined,
            editMode: true,
            diagramMode: 'action',
        };
        if (props.parseFragment) {
            FragmentUtils.setParseFragmentFn(props.parseFragment);
        }
    }

    getChildContext() {
        return {
            getEndpoints: this.props.getEndpoints,
            astRoot: this.state.currentAST,
            goToSource: this.props.goToSource,
        };
    }

    componentDidMount() {
        this.updateDiagram(this.props.docUri);
    }

    componentWillReceiveProps(nextProps) {
        this.updateDiagram(nextProps.docUri);
    }

    onModelUpdate(evt) {
        this.forceUpdate();
        const { currentAST } = this.state;
        if (currentAST) {
            this.props.onChange({ newAST: currentAST });
        }
    }

    updateDiagram(docUri) {
        this.props.getAST(docUri)
                .then((parserReply) => {
                    const { currentAST } = this.state;
                    if (parserReply.ast) {
                        if (currentAST) {
                            currentAST.off(TREE_MODIFIED, this.onModelUpdate);
                        }
                        const newAST = TreeBuilder.build(parserReply.ast);
                        if (newAST) {
                            newAST.on(TREE_MODIFIED, this.onModelUpdate.bind(this));
                        }
                        this.setState({
                            currentAST: newAST,
                        });
                    }
                });
    }

    render() {
        const { currentAST, diagramMode, editMode } = this.state;
        const { width, height } = this.props;
        if (!currentAST) {
            return (
                <Dimmer active inverted>
                    <Loader size='large'></Loader>
                </Dimmer>
            );
        }
        return (
            <React.Fragment>
                <div className='ballerina-editor design-view-container'>
                    <BallerinaDesignView
                        model={currentAST}
                        mode={diagramMode}
                        editMode={editMode}
                        height={height}
                        width={width}
                        onModeChange={(evt) => {
                            this.setState({
                                editMode: evt.editMode,
                            });
                        }}
                        onCodeExpandToggle={(evt) => {
                            this.setState({
                                diagramMode: evt.mode,
                            });
                        }}
                    />
                </div>
            </React.Fragment>);
    }
}

BallerinaDiagramWrapper.childContextTypes = {
    getEndpoints: PropTypes.func,
    astRoot: PropTypes.instanceOf(Object).isRequired,
    goToSource: PropTypes.func.isRequired,
};

BallerinaDiagramWrapper.propTypes = {
    getAST: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    docUri: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    getEndpoints: PropTypes.func,
    parseFragment: PropTypes.func,
    goToSource: PropTypes.func.isRequired,
};

export default BallerinaDiagramWrapper;
