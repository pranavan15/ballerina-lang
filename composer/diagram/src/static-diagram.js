/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

import { renderToStaticMarkup } from 'react-dom/server';
import { createElement } from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';

import Diagram from 'plugins/ballerina/diagram/diagram';
import TreeBuilder from 'plugins/ballerina/model/tree-builder';
import '@ballerina/theme/default-theme/index.js/';
import './scss/themes/default.scss';
import '@ballerina/theme/font-ballerina';

const BalDiagram = DragDropContext(HTML5Backend)(Diagram);

/**
 * Renders static Ballerina Diagram for
 * given AST on given HTML element.
 *
 * @param {HTMLElement} target Target Element to render on
 * @param {Object} modelJson Source AST
 * @param {Object} props Props for Diagram Component
 */
export default function renderStaticDiagram(target, modelJson, props = {}) {
    const defaultProps = {
        model: TreeBuilder.build(modelJson),
        mode: 'action',
        editMode: true,
        height: 300,
        width: 300,
        zoom: 1,
    };
    target.classList.add('composer-library');
    Object.assign(defaultProps, props);
    const el = createElement(BalDiagram, defaultProps);
    target.innerHTML = renderToStaticMarkup(el);
}
