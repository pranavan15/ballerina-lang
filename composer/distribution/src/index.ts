import { renderEditableDiagram, renderStaticDiagram } from "@ballerina/diagram";
import { renderSamplesList } from "@ballerina/ballerina-by-examples";

export {
    renderEditableDiagram,
    renderStaticDiagram,
    renderSamplesList
}