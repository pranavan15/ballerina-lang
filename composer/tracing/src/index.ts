export * from './TraceLogs';
export * from './DetailView';