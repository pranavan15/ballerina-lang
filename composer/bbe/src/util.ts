import { createElement } from 'react';
import { render } from 'react-dom';
import { SamplesList } from './ExampleList';
import { BallerinaExampleCategory } from './model';

export function renderSamplesList(target: HTMLElement,
    openSample: (url: string) => void,
    getSamples: () => Promise<Array<BallerinaExampleCategory>>,
    openLink: (url: string) => void) {
    const props = {
        getSamples,
        openSample,
        openLink,
    };
    target.classList.add('composer-library');
    const SamplesListElement = createElement(SamplesList, props);
    render(SamplesListElement, target);
}