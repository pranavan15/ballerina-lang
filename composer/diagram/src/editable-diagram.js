/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
import { createElement } from 'react';
import ReactDOM from 'react-dom';
import BallerinaDiagramWrapper from './BallerinaDiagramWrapper';
import '@ballerina/theme/default-theme/index.js/';
import './scss/themes/default.scss';
import '@ballerina/theme/font-ballerina';

/**
 * Renders Editable Ballerina Diagram on given HTMLElement.
 *
 * @param {HTMLElement} target Element to render on.
 * @param {string} docUri URI of the ballerina File. This will be used for subsequent API calls.
 * @param {number} width Preferred width in pixels.
 * @param {number} height Preferred Height in pixels.
 * @param {Function} getAST API to invoke to get AST given the URI.
 * @param {Function} onChange Will be invoked upon changes to Diagram. DocURI and new AST will be passed as args.
 * @param {Function} getEndpoints Will be invoked to get available endpoints.
 * @param {Function} parseFragment Will be invoked to get Ballerina Fragments parsed.
 * @param {Function} goToSource  Will be invoked to when a go-to-source button is cliked in Diagram.
 *                               Position info will be passed as args.
 *
 */
export default function renderEditableDiagram(target, docUri, width, height,
    getAST = () => Promise.resolve({}),
    onChange = () => {},
    getEndpoints,
    parseFragment,
    goToSource) {
    const props = {
        getAST,
        onChange,
        docUri,
        width,
        height,
        getEndpoints,
        parseFragment,
        goToSource,
    };
    target.classList.add('composer-library');
    const BalDiagramElement = createElement(BallerinaDiagramWrapper, props);
    ReactDOM.render(BalDiagramElement, target);
}
