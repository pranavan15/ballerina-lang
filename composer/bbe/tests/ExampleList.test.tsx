import * as React from 'react';
import { SamplesList } from '../src';
import { create } from 'react-test-renderer';
import getBBEs from './../stories/bbes';

test('SamplesList renders properly', () => {
    function openSample(url: string) {

    }
    var focusOnInputSpy = jest.fn();
    jest
        .spyOn(SamplesList.prototype, 'focusOnSearchInput')
        .mockImplementation(focusOnInputSpy);

    const component = create(
        <SamplesList getSamples={getBBEs} openSample={openSample} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(focusOnInputSpy).toHaveBeenCalled();
});