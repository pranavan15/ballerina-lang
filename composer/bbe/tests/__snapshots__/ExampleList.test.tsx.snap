// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SamplesList renders properly 1`] = `
<div
  className="ui grid welcome-page"
>
  <div
    className="two column row welcome-navbar"
  >
    <div
      className="column nav-tagline"
    >
      Search and open available examples
    </div>
    <div
      className="column"
    >
      <div
        className="top-nav-links"
        style={
          Object {
            "marginRight": 40,
            "paddingRight": 0,
          }
        }
      >
        <div
          className="ui loading icon input"
        >
          <input
            onChange={[Function]}
            placeholder="Search"
            type="text"
          />
          <i
            aria-hidden="true"
            className="spinner icon"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row welcome-content-wrapper"
  >
    <div
      className="sixteen wide computer sixteen wide mobile sixteen wide tablet column rightContainer"
    >
      <div
        className="ui grid"
      />
    </div>
  </div>
</div>
`;
